libs/custom-scanner/src/custom_a4_scanner.ts

```diff
@@ -0,0 +1,626 @@
+  ): Promise<Result<SheetOf<ImageFromScanner>, ErrorCode>> {
+    const timer = time(debug, 'scan');
+
+    timer.checkpoint('requesting lock');
+    const { unlock } = await this.publicApiMutex.asyncLock();
+    timer.checkpoint('got lock');
+
+    const { wantedScanSide: scanSide } = scanParameters;
+    const scannerImageA: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.A,
+      imageResolution: scanParameters.resolution,
+    };
+    const scannerImageB: ImageFromScanner = {
+      imageBuffer: Buffer.alloc(0),
+      imageWidth: 0,
+      imageHeight: 0,
+      imageDepth: scanParameters.imageColorDepth,
+      imageFormat: ImageFileFormat.Jpeg,
+      scanSide: ScanSide.B,
+      imageResolution: scanParameters.resolution,
+    };
+
+    timer.checkpoint('set scan parameters');
+    let setScanParametersResult = await this.setScanParametersInternal(
+      scanParameters
+    );
+
+    if (setScanParametersResult.isErr()) {
+      if (setScanParametersResult.err() === ErrorCode.JobNotValid) {
+        debug('job not valid, recreating job');
+        const createJobResult = await this.createJobInternal();
+
+        if (createJobResult.isErr()) {
+          return createJobResult;
+        }
+
+        setScanParametersResult = await this.setScanParametersInternal(
+          scanParameters
+        );
+      }
+
+      if (setScanParametersResult.isErr()) {
+        return setScanParametersResult;
+      }
+    }
+
+    // scan loop state
+    let errorCount = 0;
+    let startNoMoveNoScan = 0;
+
+    const readImageData = async (
+      currentSide: ScanSide,
+      {
+        pageSize,
+        imageWidth,
+        imageHeight,
+      }: { pageSize: number; imageWidth: number; imageHeight: number }
+    ): Promise<Result<void, ErrorCode>> => {
+      let readImageDataErrorCount = 0;
+      const scannerImage =
+        currentSide === ScanSide.A ? scannerImageA : scannerImageB;
+      for (;;) {
+        const getImagePortionBySideResult =
+          await this.getImagePortionBySideInternal(
+            currentSide,
+            pageSize,
+            imageWidth,
+            imageHeight
+          );
+
+        /* istanbul ignore next */
+        if (getImagePortionBySideResult.isErr()) {
+          readImageDataErrorCount += 1;
+          if (readImageDataErrorCount < maxRetries) {
+            await sleep(10);
+            continue;
+          }
+
+          return getImagePortionBySideResult;
+        }
+
+        scannerImage.imageBuffer = Buffer.concat([
+          scannerImage.imageBuffer,
+          getImagePortionBySideResult.ok(),
+        ]);
+        scannerImage.imageWidth = imageWidth;
+        scannerImage.imageHeight = imageHeight;
+
+        return ok();
+      }
+    };
+
+    async function processImageData(
```

Comment:
Consider combining `processImageData` and `readImageData` - the former just seems to wrap the latter. Could also potentially inline `getImagePortionBySideInternal` since this is the only place its used.

Reply:
Combined `processImageData` and `readImageData` but left `getImagePortionBySideInternal` to keep direct mutex access out of `scan`.